lib/ipam/ipam.go

```diff
@@ -676,67 +677,27 @@ func (c ipamClient) releaseIPsFromBlock(ctx context.Context, ips []net.IP, block
 			}
 		}
 
-		// Block exists - get the allocationBlock from the KVPair.
-		b := allocationBlock{obj.Value.(*model.AllocationBlock)}
+		// Determine whether or not the block's pool still matches the node.
+		released, err := c.maybeReleaseAffinity(ctx, obj.Value.(*model.AllocationBlock))
+		if err != nil {
+			return nil, err
+		}
 
-		// Retrieve node for this allocation. We do this so we can clean up affinity for blocks
-		// which should no longer be affine to this host.
-		host := getHostAffinity(b.AllocationBlock)
-		logCtx.Debugf("block affinity: %s", host)
-		if host != "" {
-			// Get the corresponding node object.
-			logCtx.Debugf("Looking up node for host affinity %s", host)
-			node, err := c.client.Get(ctx, model.ResourceKey{Kind: v3.KindNode, Name: host}, "")
+		// Since this updates the block, we need to requery it so the code below
```

Comment:
I'd tweak this comment slightly now to say "Releasing the affinity modifies the block, so we need to re-query it if it was released..."